node/network/protocol/src/lib.rs

```diff
@@ -336,6 +336,66 @@ pub mod v1 {
 		Statement(Hash, SignedFullStatement)
 	}
 
+	#[derive(Debug, Clone, Copy, PartialEq, thiserror::Error)]
+	#[allow(missing_docs)]
+	pub enum CompressedPoVError {
+		#[error("Failed to compress a PoV")]
+		Compress,
+		#[error("Failed to decompress a PoV")]
+		Decompress,
+		#[error("Failed to decode the uncompressed PoV")]
+		Decode,
+		#[error("Architecture is not supported")]
+		NotSupported,
+	}
+
+	/// SCALE and GZip encoded [`PoV`].
+	#[derive(Debug, Clone, Encode, Decode, PartialEq)]
+	pub struct CompressedPoV(Vec<u8>);
+
+	impl CompressedPoV {
+		/// Compress the given [`PoV`] and returns a [`CompressedPoV`].
+		#[cfg(not(target_os = "unknown"))]
+		pub fn compress(pov: &PoV) -> Result<Self, CompressedPoVError> {
+			zstd::encode_all(pov.encode().as_slice(), 3).map_err(|_| CompressedPoVError::Compress).map(Self)
+		}
+
+		/// Compress the given [`PoV`] and returns a [`CompressedPoV`].
+		#[cfg(target_os = "unknown")]
+		pub fn compress(_: &PoV) -> Result<Self, CompressedPoVError> {
+			Err(CompressedPoVError::NotSupported)
+		}
+
+		/// Decompress `self` and returns the [`PoV`] on success.
+		#[cfg(not(target_os = "unknown"))]
+		pub fn decompress(&self) -> Result<PoV, CompressedPoVError> {
+			use std::io::Read;
+			const MAX_POV_BLOCK_SIZE: usize = 32 * 1024 * 1024;
+
+			struct InputDecoder<'a, T: std::io::BufRead>(&'a mut zstd::Decoder<T>, usize);
+			impl<'a, T: std::io::BufRead> parity_scale_codec::Input for InputDecoder<'a, T> {
+				fn read(&mut self, into: &mut [u8]) -> Result<(), parity_scale_codec::Error> {
+					if self.1.saturating_add(into.len()) > MAX_POV_BLOCK_SIZE {
```

Comment:
where do we modify `self.1`?
can we have a test for this please?

Reply:
We modify `self.1` directly here? We call `saturating_add` in this line.

Reply:
`saturating_add` takes `self` by value, not `&mut self`
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=da4bed5cde4b3b145bc8754b705aeb20

Reply:
I'm an idiot :facepalm: